<img src = "https://velocitypartners.com/wp-content/uploads/2011/11/b2b-marketing-and-moneyball1.jpg">

## The MoneyBall Project

## Rules of Baseball

You don't need to know much about baseball to complete this exercise. if you're totally unfamiliar with baseball, you can watch this useful [explanatory video](https://www.youtube.com/watch?v=0bKkGeROiPA)

## [Breaking Biases](https://www.youtube.com/watch?v=yGf6LNWY9AI)

## The 2002 Oakland A's
The Oakland Athletics' 2002 season was the team's 35th in Oakland, California. It was also the 102nd season in franchise history. The Athletics finished first in the American League West with a record of 103-59.

The Athletics' 2002 campaign ranks among the most famous in franchise history. Following the 2001 season, Oakland saw the departure of three key players (the lost boys). Billy Beane, the team's general manager, responded with a series of under-the-radar free agent signings. The new-look Athletics, despite a comparative lack of star power, surprised the baseball world by besting the 2001 team's regular season record. The team is most famous, however, for winning 20 consecutive games between August 13 and September 4, 2002.[1] The Athletics' season was the subject of Michael Lewis' 2003 book Moneyball: The Art of Winning an Unfair Game (as Lewis was given the opportunity to follow the team around throughout that season)

This project is based off the book written by Michael Lewis (later turned into a movie).

## Moneyball Book
The central premise of book Moneyball is that the collective wisdom of baseball insiders (including players, managers, coaches, scouts, and the front office) over the past century is subjective and often flawed. Statistics such as stolen bases, runs batted in, and batting average, typically used to gauge players, are relics of a 19th-century view of the game and the statistics available at that time. The book argues that the Oakland A's' front office took advantage of more analytical gauges of player performance to field a team that could better compete against richer competitors in Major League Baseball (MLB).

Rigorous statistical analysis had demonstrated that on-base percentage and slugging percentage are better indicators of offensive success, and the A's became convinced that these qualities were cheaper to obtain on the open market than more historically valued qualities such as speed and contact. These observations often flew in the face of conventional baseball wisdom and the beliefs of many baseball scouts and executives.

By re-evaluating the strategies that produce wins on the field, the 2002 Athletics, with approximately US 44 million dollars in salary, were competitive with larger market teams such as the New York Yankees, who spent over US$125 million in payroll that same season.

<img src = "https://upload.wikimedia.org/wikipedia/commons/a/ae/MONEYBALLchart.png">

Because of the team's smaller revenues, Oakland is forced to find players undervalued by the market, and their system for finding value in undervalued players has proven itself thus far. This approach brought the A's to the playoffs in 2002 and 2003.

In this project we'll work with some data and with the goal of trying to find replacement players for the ones lost at the start of the off-season - During the 2001–02 offseason, the team lost three key free agents to larger market teams: 2000 AL MVP Jason Giambi to the New York Yankees, outfielder Johnny Damon to the Boston Red Sox, and closer Jason Isringhausen to the St. Louis Cardinals.

The main goal of this project is for you to feel comfortable working with R on real data to try and derive actionable insights!

## Let's get started!
Follow the steps outlined in bold below using your new R skills and help the Oakland A's recruit under-valued players!

<img src = "https://miro.medium.com/max/1400/1*RmFOEGRNw5oq6U5Ahy9bRw.jpeg">

## Data
We'll be using data from [Sean Lahaman's Website](http://www.seanlahman.com/baseball-archive/statistics/); a very useful source for baseball statistics. The documentation for the csv files is located in the readme2013.txt file. You may need to reference this to understand what acronyms stand for.

__Use R to open the Batting.csv file and assign it to a dataframe called batting using read.csv__

In [2]:
# Load in the Libraries we will need
library(ggplot2)
library(plotly)
library(dplyr)
library(tidyr)
library(formattable)

playerID  yearID stint teamID lgID G  G_batting AB R H ... SB CS BB SO IBB
1 aardsda01 2004   1     SFN    NL   11 11        0  0 0 ... 0  0  0  0  0  
2 aardsda01 2006   1     CHN    NL   45 43        2  0 0 ... 0  0  0  0  0  
3 aardsda01 2007   1     CHA    AL   25  2        0  0 0 ... 0  0  0  0  0  
4 aardsda01 2008   1     BOS    AL   47  5        1  0 0 ... 0  0  0  1  0  
5 aardsda01 2009   1     SEA    AL   73  3        0  0 0 ... 0  0  0  0  0  
6 aardsda01 2010   1     SEA    AL   53  4        0  0 0 ... 0  0  0  0  0  
  HBP SH SF GIDP G_old
1 0   0  0  0    11   
2 0   1  0  0    45   
3 0   0  0  0     2   
4 0   0  0  0     5   
5 0   0  0  0    NA   
6 0   0  0  0    NA

'data.frame':	97889 obs. of  24 variables:
 $ playerID : Factor w/ 18107 levels "aardsda01","aaronha01",..: 1 1 1 1 1 1 1 2 2 2 ...
 $ yearID   : int  2004 2006 2007 2008 2009 2010 2012 1954 1955 1956 ...
 $ stint    : int  1 1 1 1 1 1 1 1 1 1 ...
 $ teamID   : Factor w/ 149 levels "ALT","ANA","ARI",..: 117 35 33 16 116 116 93 80 80 80 ...
 $ lgID     : Factor w/ 6 levels "AA","AL","FL",..: 4 4 2 2 2 2 2 4 4 4 ...
 $ G        : int  11 45 25 47 73 53 1 122 153 153 ...
 $ G_batting: int  11 43 2 5 3 4 NA 122 153 153 ...
 $ AB       : int  0 2 0 1 0 0 NA 468 602 609 ...
 $ R        : int  0 0 0 0 0 0 NA 58 105 106 ...
 $ H        : int  0 0 0 0 0 0 NA 131 189 200 ...
 $ X2B      : int  0 0 0 0 0 0 NA 27 37 34 ...
 $ X3B      : int  0 0 0 0 0 0 NA 6 9 14 ...
 $ HR       : int  0 0 0 0 0 0 NA 13 27 26 ...
 $ RBI      : int  0 0 0 0 0 0 NA 69 106 92 ...
 $ SB       : int  0 0 0 0 0 0 NA 2 3 2 ...
 $ CS       : int  0 0 0 0 0 0 NA 2 1 4 ...
 $ BB       : int  0 0 0 0 0 0 NA 28 49 37 ...
 $ S

__Make sure you understand how to call the columns by using the $ symbol.__

__Call the head() of the first five rows of AB (At Bats) column__

[1] 0 2 0 1 0

__Call the head of the doubles (X2B) column__

[1] 0 0 0 0 0

## Feature Engineering
We need to add three more statistics that were used in Moneyball! These are:

- [Batting Average](https://en.wikipedia.org/wiki/Batting_average)
- [On Base Percentage](http://en.wikipedia.org/wiki/On-base_percentage)
- [Slugging Percentage](http://en.wikipedia.org/wiki/Slugging_percentage)

Click on the links provided and search the wikipedia page for the formula for creating the new statistic! 

For example, for Batting Average, you'll need to scroll down until you see:

 $$AVG=\frac{H}{AB}$$
 
Which means that the Batting Average is equal to H (Hits) divided by AB (At Base). So we'll do the following to create a new column called __BA__ and add it to our data frame:

In [8]:
batting$BA <- batting$H / batting$AB

__After doing this operation, check the last 5 entries of the BA column of your data frame and it should look like this:__

[1] 0.1230769 0.2746479 0.1470588 0.2745098 0.2138728

__Now do the same for some new columns! On Base Percentage (OBP) and Slugging Percentage (SLG). Hint: For SLG, you need 1B (Singles), this isn't in your data frame. However you can calculate it by subtracting doubles,triples, and home runs from total hits (H): 1B = H-2B-3B-HR__

- __Create an On Base Percentage (OBP) Column__
- __Create a Slugging Average (SLG) Column__

In [10]:
# On Base Percentage


# Creating X1B (Singles)


# Creating Slugging Average (SLG) 


'data.frame':	97889 obs. of  28 variables:
 $ playerID : Factor w/ 18107 levels "aardsda01","aaronha01",..: 1 1 1 1 1 1 1 2 2 2 ...
 $ yearID   : int  2004 2006 2007 2008 2009 2010 2012 1954 1955 1956 ...
 $ stint    : int  1 1 1 1 1 1 1 1 1 1 ...
 $ teamID   : Factor w/ 149 levels "ALT","ANA","ARI",..: 117 35 33 16 116 116 93 80 80 80 ...
 $ lgID     : Factor w/ 6 levels "AA","AL","FL",..: 4 4 2 2 2 2 2 4 4 4 ...
 $ G        : int  11 45 25 47 73 53 1 122 153 153 ...
 $ G_batting: int  11 43 2 5 3 4 NA 122 153 153 ...
 $ AB       : int  0 2 0 1 0 0 NA 468 602 609 ...
 $ R        : int  0 0 0 0 0 0 NA 58 105 106 ...
 $ H        : int  0 0 0 0 0 0 NA 131 189 200 ...
 $ X2B      : int  0 0 0 0 0 0 NA 27 37 34 ...
 $ X3B      : int  0 0 0 0 0 0 NA 6 9 14 ...
 $ HR       : int  0 0 0 0 0 0 NA 13 27 26 ...
 $ RBI      : int  0 0 0 0 0 0 NA 69 106 92 ...
 $ SB       : int  0 0 0 0 0 0 NA 2 3 2 ...
 $ CS       : int  0 0 0 0 0 0 NA 2 1 4 ...
 $ BB       : int  0 0 0 0 0 0 NA 28 49 37 ...
 $ S

      playerID         yearID         stint           teamID        lgID      
 mcguide01:   31   Min.   :1871   Min.   :1.000   CHN    : 4720   AA  : 1890  
 henderi01:   29   1st Qu.:1931   1st Qu.:1.000   PHI    : 4621   AL  :44369  
 newsobo01:   29   Median :1970   Median :1.000   PIT    : 4575   FL  :  470  
 johnto01 :   28   Mean   :1962   Mean   :1.077   SLN    : 4535   NL  :49944  
 kaatji01 :   28   3rd Qu.:1995   3rd Qu.:1.000   CIN    : 4393   PL  :  147  
 ansonca01:   27   Max.   :2013   Max.   :5.000   CLE    : 4318   UA  :  332  
 (Other)  :97717                                  (Other):70727   NA's:  737  
       G            G_batting            AB              R         
 Min.   :  1.00   Min.   :  0.00   Min.   :  0.0   Min.   :  0.00  
 1st Qu.: 13.00   1st Qu.:  7.00   1st Qu.:  9.0   1st Qu.:  0.00  
 Median : 35.00   Median : 32.00   Median : 61.0   Median :  5.00  
 Mean   : 51.65   Mean   : 49.13   Mean   :154.1   Mean   : 20.47  
 3rd Qu.: 81.00   3rd Qu.: 8

## Merging Salary Data with Batting Data
We know we don't just want the best players, we want the most undervalued players, meaning we will also need to know current salary information! We have salary information in the csv file 'Salaries.csv'.

Complete the following steps to merge the salary data with the player stats!

__Load the Salaries.csv file into a dataframe called sal using read.csv__

yearID teamID lgID playerID  salary 
1 1985   BAL    AL   murraed02 1472819
2 1985   BAL    AL   lynnfr01  1090000
3 1985   BAL    AL   ripkeca01  800000
4 1985   BAL    AL   lacyle01   725000
5 1985   BAL    AL   flanami01  641667
6 1985   BAL    AL   boddimi01  625000

__Use summary to get a summary of the batting data frame and notice the minimum year in the yearID column. Our batting data goes back to 1871! Our salary data starts at 1985, meaning we need to remove the batting data that occured before 1985.__

__Use subset() to reassign batting to only contain data from 1985 and onwards__

      playerID         yearID         stint          teamID      lgID      
 moyerja01:   27   Min.   :1985   Min.   :1.00   SDN    : 1313   AA:    0  
 mulhote01:   26   1st Qu.:1993   1st Qu.:1.00   CLE    : 1306   AL:17226  
 weathda01:   26   Median :2000   Median :1.00   PIT    : 1299   FL:    0  
 maddugr01:   25   Mean   :2000   Mean   :1.08   NYN    : 1297   NL:18426  
 sierrru01:   25   3rd Qu.:2007   3rd Qu.:1.00   BOS    : 1279   PL:    0  
 thomeji01:   25   Max.   :2013   Max.   :4.00   CIN    : 1279   UA:    0  
 (Other)  :35498                                 (Other):27879             
       G           G_batting            AB              R         
 Min.   :  1.0   Min.   :  0.00   Min.   :  0.0   Min.   :  0.00  
 1st Qu.: 14.0   1st Qu.:  4.00   1st Qu.:  3.0   1st Qu.:  0.00  
 Median : 34.0   Median : 27.00   Median : 47.0   Median :  4.00  
 Mean   : 51.7   Mean   : 46.28   Mean   :144.7   Mean   : 19.44  
 3rd Qu.: 77.0   3rd Qu.: 77.00   3rd Qu.:241.0   3rd Qu.

Now it is time to merge the batting data with the salary data! Since we have players playing multiple years, we'll have repetitions of playerIDs for multiple years, meaning we want to merge on both players and years.

__Use the merge() function to merge the batting and sal data frames by c('playerID','yearID'). Call the new data frame "combo"__

      playerID         yearID         stint          teamID.x     lgID.x    
 moyerja01:   27   Min.   :1985   Min.   :1.000   LAN    :  940   AA:    0  
 thomeji01:   25   1st Qu.:1993   1st Qu.:1.000   PHI    :  937   AL:12292  
 weathda01:   25   Median :1999   Median :1.000   BOS    :  935   FL:    0  
 vizquom01:   24   Mean   :1999   Mean   :1.098   NYA    :  928   NL:13105  
 gaettga01:   23   3rd Qu.:2006   3rd Qu.:1.000   CLE    :  920   PL:    0  
 griffke02:   23   Max.   :2013   Max.   :4.000   SDN    :  914   UA:    0  
 (Other)  :25250                                  (Other):19823             
       G            G_batting            AB              R         
 Min.   :  1.00   Min.   :  0.00   Min.   :  0.0   Min.   :  0.00  
 1st Qu.: 26.00   1st Qu.:  8.00   1st Qu.:  5.0   1st Qu.:  0.00  
 Median : 50.00   Median : 42.00   Median : 85.0   Median :  9.00  
 Mean   : 64.06   Mean   : 57.58   Mean   :182.4   Mean   : 24.71  
 3rd Qu.:101.00   3rd Qu.:101.00   3rd Qu.:3

## Analyzing the Lost Players
As previously mentioned, the Oakland A's lost 3 key players during the off-season. We'll want to get their stats to see what we have to replace. The players lost were: 
- Jason Giambi (giambja01) 
    - First Baseman 
    - 2000 American League MVP 
    - Lost to the New York Yankees 
- Johnny Damon (damonjo01)
    - Outfielder 
    - Lost to the Boston Red Sox 
- Rainer Gustavo "Ray" Olmedo ('saenzol01')
    - Infielder
    - Lost to the Los Angeles Dodgers
    
__Use the subset() function to get a data frame called lost_players from the combo data frame consisting of those 3 players. Hint: Try to figure out how to use %in% to avoid a bunch of or statements!__

playerID  yearID stint teamID.x lgID.x G   G_batting AB  R   H   ... SF
5135  damonjo01 1995   1     KCA      AL      47  47       188  32  53 ...  3
5136  damonjo01 1996   1     KCA      AL     145 145       517  61 140 ...  5
5137  damonjo01 1997   1     KCA      AL     146 146       472  70 130 ...  1
5138  damonjo01 1998   1     KCA      AL     161 161       642 104 178 ...  3
5139  damonjo01 1999   1     KCA      AL     145 145       583 101 179 ...  4
5140  damonjo01 2000   1     KCA      AL     159 159       655 136 214 ... 12
5141  damonjo01 2001   1     OAK      AL     155 155       644 108 165 ...  4
5142  damonjo01 2002   1     BOS      AL     154 154       623 118 178 ...  5
5143  damonjo01 2003   1     BOS      AL     145 145       608 103 166 ...  6
5144  damonjo01 2004   1     BOS      AL     150 150       621 123 189 ...  3
5145  damonjo01 2005   1     BOS      AL     148 148       624 117 197 ...  9
5146  damonjo01 2006   1     NYA      AL     149 149       593 115 169 ...  5
5147  damonjo01 2007   1     NYA      AL     141 141       533  93 144 ...  3
5148  damonjo01 2008   1     NYA      AL     143 143       555  95 168 ...  1
5149  damonjo01 2009   1     NYA      AL     143 143       550 107 155 ...  1
5150  damonjo01 2010   1     DET      AL     145 145       539  81 146 ...  1
5151  damonjo01 2011   1     TBA      AL     150 150       582  79 152 ...  5
7872  giambja01 1995   1     OAK      AL      54  54       176  27  45 ...  2
7873  giambja01 1996   1     OAK      AL     140 140       536  84 156 ...  5
7874  giambja01 1997   1     OAK      AL     142 142       519  66 152 ...  8
7875  giambja01 1998   1     OAK      AL     153 153       562  92 166 ...  9
7876  giambja01 1999   1     OAK      AL     158 158       575 115 181 ...  8
7877  giambja01 2000   1     OAK      AL     152 152       510 108 170 ...  8
7878  giambja01 2001   1     OAK      AL     154 154       520 109 178 ...  9
7879  giambja01 2002   1     NYA      AL     155 155       560 120 176 ...  5
7880  giambja01 2003   1     NYA      AL     156 156       535  97 134 ...  5
7881  giambja01 2004   1     NYA      AL      80  80       264  33  55 ...  3
7882  giambja01 2005   1     NYA      AL     139 139       417  74 113 ...  1
7883  giambja01 2006   1     NYA      AL     139 139       446  92 113 ...  7
7884  giambja01 2007   1     NYA      AL      83  83       254  31  60 ...  1
7885  giambja01 2008   1     NYA      AL     145 145       458  68 113 ...  9
7886  giambja01 2009   1     OAK      AL      83  83       269  39  52 ...  2
7887  giambja01 2009   2     COL      NL      19  19        24   4   7 ...  0
7888  giambja01 2010   1     COL      NL      87  87       176  17  43 ...  5
7889  giambja01 2011   1     COL      NL      64  64       131  20  34 ...  1
7890  giambja01 2012   1     COL      NL      60  NA        89   7  20 ...  2
7891  giambja01 2013   1     CLE      AL      71  71       186  21  34 ...  3
20112 saenzol01 1999   1     OAK      AL      97  97       255  41  70 ...  3
20113 saenzol01 2000   1     OAK      AL      76  76       214  40  67 ...  1
20114 saenzol01 2001   1     OAK      AL     106 106       305  33  67 ...  3
20115 saenzol01 2002   1     OAK      AL      68  68       156  15  43 ...  2
20116 saenzol01 2005   1     LAN      NL     109 109       319  39  84 ...  2
20117 saenzol01 2006   1     LAN      NL     103 103       179  30  53 ...  4
20118 saenzol01 2007   1     LAN      NL      92  92       110   9  21 ...  4
      GIDP G_old BA        OBP       X1B SLG       teamID.y lgID.y salary  
5135   2    47   0.2819149 0.3235294  34 0.4414894 KCA      AL       109000
5136   4   145   0.2707930 0.3129496 107 0.3675048 KCA      AL       180000
5137   3   146   0.2754237 0.3378378 102 0.3855932 KCA      AL       240000
5138   4   161   0.2772586 0.3394625 120 0.4392523 KCA      AL       460000
5139  13   145   0.3070326 0.3789954 117 0.4768439 KCA      AL      2100000
5140   7   159   0.3267176 0.3819918 146

      playerID      yearID         stint          teamID.x  lgID.x 
 giambja01:20   Min.   :1995   Min.   :1.000   OAK    :13   AA: 0  
 damonjo01:17   1st Qu.:2000   1st Qu.:1.000   NYA    :11   AL:37  
 saenzol01: 7   Median :2004   Median :1.000   KCA    : 6   FL: 0  
 aardsda01: 0   Mean   :2004   Mean   :1.023   BOS    : 4   NL: 7  
 aaronha01: 0   3rd Qu.:2007   3rd Qu.:1.000   COL    : 4   PL: 0  
 aaronto01: 0   Max.   :2013   Max.   :2.000   LAN    : 3   UA: 0  
 (Other)  : 0                                  (Other): 3          
       G           G_batting           AB              R         
 Min.   : 19.0   Min.   : 19.0   Min.   : 24.0   Min.   :  4.00  
 1st Qu.: 86.0   1st Qu.: 89.5   1st Qu.:244.0   1st Qu.: 32.75  
 Median :142.5   Median :143.0   Median :513.5   Median : 76.50  
 Mean   :120.7   Mean   :122.1   Mean   :415.3   Mean   : 69.86  
 3rd Qu.:150.0   3rd Qu.:150.0   3rd Qu.:565.2   3rd Qu.:104.75  
 Max.   :161.0   Max.   :161.0   Max.   :655.0   Max.   :136

__Since all these players were lost in after 2001 in the offseason, let's only concern ourselves with the data from 2001.__

__Use subset again to only grab the rows where the yearID was 2001.__

__Reduce the lost_players data frame to the following columns: playerID,H,X2B,X3B,HR,OBP,SLG,BA,AB__

playerID  H   X2B X3B HR OBP       SLG       BA        AB  salary 
5141  damonjo01 165 34  4    9 0.3235294 0.3633540 0.2562112 644 7100000
7878  giambja01 178 47  2   38 0.4769001 0.6596154 0.3423077 520 4103333
20114 saenzol01  67 21  1    9 0.2911765 0.3836066 0.2196721 305  290000

## Replacement Players
Now we have all the information we need! Here is your final task - Find Replacement Players for the key three players we lost! However, you have three constraints:

- The total combined salary of the three players can not exceed 15 million dollars.
- Their combined number of At Bats (AB) needs to be equal to or greater than the lost players.
- Their mean OBP had to equal to or greater than the mean OBP of the lost players

Use the combo dataframe you previously created as the source of information! Remember to just use the 2001 subset of that dataframe. There's lost of different ways you can do this, so be creative! It should be relatively simple to find 3 players that satisfy the requirements, note that there are many correct combinations available!

